geonode/harvesting/harvesters/base.py

```diff
@@ -81,13 +85,38 @@ def check_availability(self, timeout_seconds: typing.Optional[int] = 5) -> bool:
         """Check whether the remote service is online"""
 
     @abc.abstractmethod
-    def get_resource(
+    def get_resource_type_class(self, resource_type: str):
+        """ Return resource type class from resource type string """
+
+    @abc.abstractmethod
+    def get_layer_resource(
             self,
-            resource_unique_identifier: str,
-            resource_type: str,
-            harvesting_session_id: typing.Optional[int] = None
-    ) -> typing.Optional[resourcedescriptor.RecordDescription]:
-        """Harvest a single resource from the remote service"""
+            resource_unique_identifier: str
+    ) -> typing.Optional[resourcedescriptor.LayerData]:
+        """
+        Harvest a single layer resource from the remote service.
+        Return None if it does not have this resource.
+        """
+
+    @abc.abstractmethod
+    def get_map_resource(
+            self,
+            resource_unique_identifier: str
+    ) -> typing.Optional[resourcedescriptor.MapData]:
+        """
+        Harvest a single map resource from the remote service.
+        Return None if it does not have this resource.
+        """
+
+    @abc.abstractmethod
+    def get_document_resource(
+            self,
+            resource_unique_identifier: str
+    ) -> typing.Optional[resourcedescriptor.DocumentData]:
+        """
+        Harvest a single document resource from the remote service.
+        Return None if it does not have this resource.
+        """
```

Comment:
Replacing `BaseHarvesterWorker.get_resource()` with three additional methods (`get_layer_resource()`, `get_map_resource()` and `get_document_resource()` seems unnecessary and will make the code more bloated.

Remember that abstract methods MUST be implemented by child classes. This will force a child class that only supports harvesting of layers (e.g the WMS harvester) to implement the other methods, even if they are empty, which seems wasteful. Please revert this change back to the original implementation.

If you prefer to have separate methods for retrieving specific resource types, then those should be implemented in the child class only, and then `get_resource()` would call them.

Do not force authors of future harvester workers to have to implement empty methods

Reply:
Updated @ricardogsilva